MH-12647 MH-12756 extend external api

krinnewitz opened the pull request:
This pull request consists of two commits, both extending the external API:

# MH-12647 Add Scheduling Capabilities to External API

This patch adds the possibility to schedule events through the external
API's events endpoint. Scheduling can be done by specifying schedule
information when creating an event. Additionally, existing scheduled
events can be re-scheduled by updating the scheduling information with
new details. The following properties can be used:

- start date
- end date
- rrule (when creating new events)
- agent id
- agent inputs

Available capture agents can be retrieved from the new /agents endpoint
which is also added by this commit.

See the documentation for more details about scheduling events or
querying capture agents.

This also increases the external API version to v1.1.0 since new fields
were added.

# MH-12756 Add Fields to External API

This patch adds missing fields to series and the events endpoints
starting API Version 1.1.0.
API Version 1.0.0 will not provide these fields.

GET api/events
and GET /api/events/{event_id}
- language
- rightsholder
- license
- is_part_of
- duration
- source

GET /api/series
and GET /api/series/{series_id}
- language
- rightsholder
- license
---
The motivation to have just one PR for both tickets is to avoid conflicts due to the need to increase the API version to 1.1.0 for both patches.

This work is sponsored by SWITCH.

Commit:
MH-12647 Add Scheduling Capabilities to External API

This patch adds the possibility to schedule events through the external
API's events endpoint. Scheduling can be done by specifying schedule
information when creating an event. Additionally, existing scheduled
events can be re-scheduled by updating the scheduling information with
new details. The following properties can be used:

- start date
- end date
- rrule (when creating new events)
- agent id
- agent inputs

Available capture agents can be retrieved from the new /agents endpoint
which is also added by this commit.

See the documentation for more details about scheduling events or
querying capture agents.

This also increases the external API version to v1.1.0 since new fields
were added.

doofy committed:
MH-12756 add fields to external api

This patch adds missing fields to series and the events endpoints
starting API Version 1.1.0.
API Version 1.0.0 will not provide these fields.

GET api/events
and GET /api/events/{event_id}
- language
- rightsholder
- license
- is_part_of
- duration
- source

GET /api/series
and GET /api/series/{series_id}
- language
- rightsholder
- license

slampunk commented:
Good stuff, looking forward to this!

staubesv commented on modules/external-api/src/main/java/org/opencastproject/external/common/ApiVersion.java:
```diff
@@ -20,30 +20,63 @@
   private String versionString;
+  private int major;
+  private int minor;
+  private int micro;
```
Please use the name "patch" instead of "micro" to be compliant to https://semver.org/

krinnewitz replied:
good point, fixed

staubesv commented on modules/external-api/src/main/java/org/opencastproject/external/common/ApiFormat.java:
```diff
@@ -21,5 +21,5 @@
 
 public enum ApiFormat {
-  JSON, XML
+  JSON, XML, UNSPECIFIED
```
What is the reasoning behind UNSPECIFIED? It is not a supported format, is it?

krinnewitz replied:
This is used when clients send "Accept: application/*" which basically means " I accept any format". This would lead to a server error, otherwise. It happens during the unit tests.

krinnewitz replied:
I removed UNSPECIFIED as well as the unused XML value.

staubesv commented on modules/external-api/src/main/java/org/opencastproject/external/common/ApiMediaType.java:
```diff
@@ -54,6 +54,8 @@ private static ApiFormat extractFormat(String mediaType) {
     final String format;
     if (subtype.contains("+")) {
       format = subtype.substring(subtype.indexOf("+") + 1);
+    } else if ("*".equals(subtype)) {
```
What is the reasoning behind this?

krinnewitz replied:
See the comment above

krinnewitz replied:
hi @staubesv, I just thought about it again. Maybe it is better to just set JSON here and don't use UNSPECIFIED. Or even removing the whole ApiFormat enum since we only support JSON. What do you think?

krinnewitz replied:
changed to fall back to JSON

staubesv commented on modules/external-api/src/main/java/org/opencastproject/external/common/ApiVersion.java:
```diff
@@ -20,30 +20,63 @@
   }
 
+  public boolean isSmallerThan(ApiVersion other) {
+    if (this.major < other.major) return true;
```
Personally, I would prefer...

```
return ((this.major < other.major) ||
       ((this.major == other.major) && (this.minor < other.minor)) ||
       ((this.major == other.major) && (this.minor == other.minor) && (this.patch < other.patch)));
```

... but this might be a matter of taste.

staubesv commented on modules/external-api/src/main/java/org/opencastproject/external/endpoint/CaptureAgentsEndpoint.java:
```diff
@@ -0,0 +1,157 @@
+  public Response getAgent(
+      @HeaderParam("Accept") String acceptHeader,
+      @PathParam("agentId") String id) throws Exception {
+    final ApiVersion requestedVersion = ApiMediaType.parse(acceptHeader).getResponseVersion();
```
As this line appear so often, it might be worth thinking about overloading ApiResponses.Json.ok to also directly support the accetHeader as formal parameter

krinnewitz replied:
ok, done

staubesv commented on modules/external-api/src/main/java/org/opencastproject/external/common/ApiVersion.java:
```diff
@@ -20,30 +20,63 @@
 
   ApiVersion(String versionString) {
     this.versionString = versionString;
+    if ("v*.*.*".equals(versionString)) {
```
What is this useful for?

krinnewitz replied:
The constructor is called with `"v*.*.*"` when the `VERSION_UNDEFINED` enum value is instantiated.

staubesv commented on modules/external-api/src/main/java/org/opencastproject/external/endpoint/EventsEndpoint.java:
```diff
@@ -716,6 +793,14 @@ protected JValue eventToJSON(Event event, Boolean withAcl, Boolean withMetadata,
+      fields.add(f("language", v(event.getLanguage(), BLANK)));
+      fields.add(f("rightsholder", v(event.getRights(), BLANK)));
+      fields.add(f("license", v(event.getLicense(), BLANK)));
+      fields.add(f("is_part_of", v(event.getSeriesId(), BLANK)));
```
@krinnewitz Ooops, seems I did an error in the specification in MH-12756: I think the name should be isPartOf to be consistent with the /metadata request. I will double-check all field names and report back later (no need for action on your side)

gregorydlogan replied:
Ping, checking on this.  Is "is_part_of" correct?

staubesv replied:
From my point of view: Yes. See the fieldnames.pdf in https://opencast.jira.com/browse/MH-12808 to get an idea of the current state of the field names (and potential future improvements).

staubesv committed:
Various Fixes

- Adjust security configuration for new External API endpoints
- Fix review findings

lkiesow commented:
@krinnewitz, can you please fix the build error? There seem to be trailing spaces in some of the configuration files.

Commit:
Remove unused code

staubesv commented:
@krinnewitz & @lkiesow Oh, that must have been my additions. We provide a PR to Kim's repo today.

krinnewitz commented:
Hi @staubesv, a PR won't be necessary as I am currently fixing this along with your review findings.

staubesv commented on modules/external-api/src/main/java/org/opencastproject/external/common/ApiResponses.java:
```diff
@@ -82,9 +82,25 @@ public static Response serverError(String message, Object... args) {
      * @return The new {@link Response}
      */
     public static Response ok(ApiVersion version, String body) {
+      // FIXME: This does not make sense. Sending content type 'application/json' for a plain string?
```
Well, actually it should not return plain strings at all since it "officially" only accepts application/json, application/v1.0.0+json and application/v1.1.0+json...

That is, however, not scope of this PR. I've created https://opencast.jira.com/browse/MH-12798 for this. Could you please mention MH-12798 in the comment?

Maybe something like: "// MH-12798: The External API should consistently return JSON strings"

krinnewitz replied:
ok, done

staubesv committed:
Fix handling of External API formats

staubesv committed:
Add fields 'series' and 'status' to /api/events and /api/events/{id}

staubesv committed:
Add field 'description' to /api/series

krinnewitz commented on modules/external-api/src/main/java/org/opencastproject/external/util/SchedulingUtils.java:
```diff
@@ -0,0 +1,395 @@
+      final Agent agent = agentStateService.getAgent(schedulingInfo.getAgentId().get());
+      String timezone = agent.getConfiguration().getProperty("capture.device.timezone");
+      if (StringUtils.isBlank(timezone)) {
+        timezone = TimeZone.getDefault().getID();
```
This code is taken from the SchedulerService which turns out to be buggy. When the SchedulerService is fixed, we should update this code as well.

staubesv committed:
Document fields of External API 1.1.0

staubesv committed:
Fix creation of multiple scheduled events

staubesv committed:
Make REST docs work with External API

gregorydlogan commented on docs/guides/developer/docs/api/agents-api.md:
```diff
@@ -0,0 +1,78 @@
+
+Query String Parameter | Type                        | Description
+:----------------------|:----------------------------|:-----------
+`limit`                | [`integer`](types.md#basic) | The maximum number of results to return for a single request
```
These should be changed over to match the limit and offset descriptions in the other apis (see events-api.md for an example)

staubesv replied:
Fixed once https://github.com/plapadoo/opencast/pull/19 is merged.

staubesv replied:
Fixed once https://github.com/plapadoo/opencast/pull/19 is merged

gregorydlogan commented on docs/guides/developer/docs/api/events-api.md:
```diff
@@ -81,6 +81,14 @@ Field                | Type                                 | Description
 `has_previews`       | [`boolean`](types.md#basic)          | Whether this event can be opened with the video editor
 `location`\*         | [`string`](types.md#basic)           | The bibliographic location of this event
 `publication_status` | [`array[string]`](types.md#array)    | The publications available for this event
+`language`\*         | [`string`](types.md#basic)           | The language of this event (version 1.1.0 and higher)
```
Some of these have restricted inputs in the admin UI.  Are those restrictions honoured here?  Do we care if they are not?  I'm not familiar enough with the guts of DC, but if someone sets "asdfasdfasdf" as the license, is that a valid license name?

staubesv replied:
This is a GET request so it cannot be used to write any of the fields.

AFAIK, /api/events/{id}/metadata will respect the metadata catalog configuration. But we haven't changed anything in this PR related to writing metadata catalogs.

gregorydlogan commented on docs/guides/developer/docs/api/events-api.md:
```diff
@@ -151,14 +159,47 @@ __Example__
+`start`    | Yes      | [`datetime`](types.md#date-and-time) | The date and time when the recording shall start
+`end`      | Yes\*    | [`datetime`](types.md#date-and-time) | The date and time when the recording shall end
+`duration` | Yes\*    | [`integer`](types.md#basic)          | The duration of the recording in milliseconds
+`rrule`    | No       | [`string`](types.md#string)          | The optional recurrence rule used to create multiple scheduled events
```
I would be tempted to link to the rrule docs here as well.

staubesv replied:
There is a reason that you don't find a link to the rrule docs: I have really no idea whether this works correctly. But if the specification links the rrule docs, that would become part of the specification.

I suggest to not link it as nobody ever tested whether it really would work correctly. Let's focus on the minimum here which is known to work.

Edit: Just realised that the rrule specs is linked below ;-)

gregorydlogan replied:
Exactly.  I would copy that link here as well, and make it clear that this should be a string.  Something like 

```
[`rrule`](link) as a [`string`](link)
```

staubesv replied:
Ok, will do

staubesv replied:
https://github.com/plapadoo/opencast/pull/20 introduces the data type "rrule". Note that all parameters (except files) are passed as strings, so no need to clarify this.

gregorydlogan commented on docs/guides/developer/docs/api/events-api.md:
```diff
@@ -151,14 +159,47 @@ __Example__
+`duration` | Yes\*    | [`integer`](types.md#basic)          | The duration of the recording in milliseconds
+`rrule`    | No       | [`string`](types.md#string)          | The optional recurrence rule used to create multiple scheduled events
+
+\* Both `end` and `duration` must be specified for creating multiple scheduled events, either `end` or `duration` must
```
Why do they need both `end` and `duration`?  I know the answer, but someone integrating might not.  We should also make it clear what happens if `end` coincides with the start of the next capture.  For example, if a recurrence ends on Tuesday at 1000, and captures start every weekday at 1000, does it capture on that Tuesday?

staubesv replied:
This line is just to indicate whether fields are mandatory or not which in this case "somewhat depends". The meaning of start, end and duration are explained on lines 185-202. Do we need to add something here?

Not sure whether we need to add descriptions of what happens if end and start coincide as this is covered by responding with conflicts.

Shall we document the current behaviour of the implementation here?

gregorydlogan replied:
Documenting the current implementation makes sense I think.  It would also make it clear *why* we're using the fields the way we are.  I can see a naive implementation supplying both `end` and `duration` of a single event (ie start: 1100, end: 1200, duration: 3600 seconds) and then ending up confused with either an error message or invalid scheduling.

staubesv replied:
Ok, will provide more explanation in the documentation.

staubesv replied:
https://github.com/plapadoo/opencast/pull/20 is now more precise considering the use of end vs duration.

I've tested that coinciding start and end would cause a scheduling conflict, added the documentation but removed it again.

Reason:
The scope of the External API considering scheduling is to provide a facility to ask Opencast to create scheduled events and report back whether Opencast did create those or not. In the latter case, the response is supposed to indicate the reason of not creating them (e.g. a scheduling conflict).

If someone would document the scheduling conflict detection that documentation should rather go to the documentation of the Scheduler service as this is not part of the responsibility of the External API.

gregorydlogan commented on docs/guides/developer/docs/api/events-api.md:
```diff
@@ -151,14 +159,47 @@ __Example__
+To create a single scheduled event, omit the field `rrule`. The start date and time of the recording is determined by
+field `start`. The end date and time of the recording is determined by either the field `end` or `duration`.
+
+The create multiple scheduled events, the field `rrule` is used. This field contains the recurrence rule used to
```
*To* create multiple...

staubesv replied:
Fixed once https://github.com/plapadoo/opencast/pull/19 is merged

staubesv replied:
Fixed once https://github.com/plapadoo/opencast/pull/19 is merged

gregorydlogan commented on docs/guides/developer/docs/api/events-api.md:
```diff
@@ -151,14 +159,47 @@ __Example__
+Field      | Required | Type                                 | Description
+:----------|:---------|:-------------------------------------|:-----------
+`agent_id` | Yes      | [`string`](types.md#basic)           | The technical identifier of the capture agent
+`start`    | Yes      | [`datetime`](types.md#date-and-time) | The date and time when the recording shall start
```
Oh boy, here comes the fun part: Which timezone are start and end in?  What if they're in different effective TZs (ie, straddling the DST switchover)?  IMO they should *always* be in UTC, but they don't absolutely have to be...

staubesv replied:
In theory, ISO 8601 is quite clear about how the times are specified and at the end of the day, it is the backend's problem to correctly process the dates (regarding TZ and DST).

In practise, I tend to specify that UTC must be used.

What do you think?

gregorydlogan replied:
As I'm thinking about the CA API, I'm thinking a hard requirement on ISO 8601.  There's no reason for us to voluntarily take on the complexity if we don't have to.  In which case this just needs documentation to use the appropriate standard.

staubesv replied:
https://github.com/plapadoo/opencast/pull/20 restricts start and end to UTC

gregorydlogan commented on etc/security/mh_default_org.xml:
```diff
@@ -175,6 +177,7 @@
     <sec:intercept-url pattern="/api/events/*/metadata/*" method="GET" access="ROLE_ADMIN, ROLE_API_EVENTS_METADATA_VIEW"/>
     <sec:intercept-url pattern="/api/events/*/publications" method="GET" access="ROLE_ADMIN, ROLE_API_EVENTS_PUBLICATIONS_VIEW"/>
     <sec:intercept-url pattern="/api/events/*/publications/*" method="GET" access="ROLE_ADMIN, ROLE_API_EVENTS_PUBLICATIONS_VIEW"/>
+    <sec:intercept-url pattern="/api/events/*/scheduling" method="GET" access="ROLE_ADMIN, ROLE_API_EVENTS_SCHEDULING_VIEW"/>
```
Missing `ROLE_API` deliberately?

staubesv replied:
No, this is correct: ROLE_**API**_EVENTS_SCHEDULING_VIEW is used to control GET access to this particular endpoint.

gregorydlogan commented on etc/security/mh_default_org.xml:
```diff
@@ -193,6 +196,7 @@
     <sec:intercept-url pattern="/api/events/*/comments/*" method="PUT" access="ROLE_ADMIN, ROLE_API_EVENTS_COMMENTS_EDIT"/>
     <sec:intercept-url pattern="/api/events/*/metadata" method="PUT" access="ROLE_ADMIN, ROLE_API_EVENTS_METADATA_EDIT"/>
     <sec:intercept-url pattern="/api/events/*/metadata/*" method="PUT" access="ROLE_ADMIN, ROLE_API_EVENTS_METADATA_EDIT"/>
+    <sec:intercept-url pattern="/api/events/*/scheduling" method="PUT" access="ROLE_ADMIN, ROLE_API_EVENTS_SCHEDULING_EDIT"/>
```
I'm assuming in this case missing `ROLE_API` makes sense since it's a PUT/POST rather than GET.

staubesv replied:
No, this is correct: ROLE_**API**_EVENTS_SCHEDUILING_EDIT is used to control PUT access to this particular endpoint.

gregorydlogan commented on etc/security/security_sample_ldap.xml-example:
```diff
@@ -175,6 +177,7 @@
     <sec:intercept-url pattern="/api/events/*/metadata/*" method="GET" access="ROLE_ADMIN, ROLE_API_EVENTS_METADATA_VIEW"/>
     <sec:intercept-url pattern="/api/events/*/publications" method="GET" access="ROLE_ADMIN, ROLE_API_EVENTS_PUBLICATIONS_VIEW"/>
     <sec:intercept-url pattern="/api/events/*/publications/*" method="GET" access="ROLE_ADMIN, ROLE_API_EVENTS_PUBLICATIONS_VIEW"/>
+    <sec:intercept-url pattern="/api/events/*/scheduling" method="GET" access="ROLE_ADMIN, ROLE_API_EVENTS_SCHEDULING_VIEW"/>
```
Same as above

staubesv replied:
Same as above, too ;-)

gregorydlogan commented on modules/external-api/src/main/java/org/opencastproject/external/common/ApiMediaType.java:
```diff
@@ -20,57 +20,38 @@
     this.externalForm = externalForm;
   }
 
+  public static ApiMediaType parse(String acceptHeader) throws ApiMediaTypeException {
```
I'm not sure we're maintaining true backwards compatibility if we're dropping support for XML.  I'm *fine* with that, since I doubt many people used it, but this kind of change should be kept to a minimum.

staubesv replied:
The @Produces annotation so far only allowed "application/json" and "application/v1.0.0+json". With other words: It was never possible to ask for XML (and I doubt that it would have ever worked).

Therefore, this does not break compatibility.

gregorydlogan commented on modules/external-api/src/main/java/org/opencastproject/external/common/ApiVersion.java:
```diff
@@ -20,28 +20,34 @@
-  public static final ApiVersion CURRENT_VERSION = VERSION_1_0_0;
+  private int major;
+  private int minor;
+  private int patch;
```
I thought we were moving away from patches to just major.minor?  I'm fine if we keep the patch number, maybe I'm just out of the loop.

staubesv replied:
We had quite some discussions about how we proceed considering versioning as the versioning was not working in means of we could not return the API version in the responses (which again the specification says to do so).

Finally, we found that it can be fixed quite easily.

Except of this, this PR is not about API versioning at all and does not attempt to change it.

gregorydlogan commented on modules/external-api/src/main/java/org/opencastproject/external/endpoint/BaseEndpoint.java:
```diff
@@ -80,6 +80,7 @@
  * supported API.
  */
 @Path("/")
+@Produces({ "application/json", "application/v1.0.0+json", "application/v1.1.0+json" })
```
Should these be the constants defined in ApiMediaType.java?

staubesv replied:
Although not in scope of this PR: Fixed once https://github.com/plapadoo/opencast/pull/19 is merged

gregorydlogan commented on modules/external-api/src/main/java/org/opencastproject/external/endpoint/BaseEndpoint.java:
```diff
@@ -199,19 +195,19 @@ public Response getOrganizationProperties() {
   @GET
   @Path("version/default")
-  @Produces({ "application/json", "application/v1.0.0+json" })
+  @Produces({ "application/json", "application/v1.0.0+json", "application/v1.1.0+json" })
```
Constants?

staubesv replied:
Fixed once https://github.com/plapadoo/opencast/pull/19 is merged

gregorydlogan commented on modules/external-api/src/main/java/org/opencastproject/external/endpoint/CaptureAgentsEndpoint.java:
```diff
@@ -0,0 +1,153 @@
+import javax.ws.rs.core.Response;
+
+@Path("/")
+@Produces({ "application/json", "application/v1.1.0+json" })
```
Constants?

staubesv replied:
Fixed once https://github.com/plapadoo/opencast/pull/19 is merged

gregorydlogan commented on modules/external-api/src/main/java/org/opencastproject/external/endpoint/CaptureAgentsEndpoint.java:
```diff
@@ -0,0 +1,153 @@
+      description = "Returns a list of agents.",
+      returnDescription = "",
+      restParameters = {
+          @RestParameter(name = "limit", description = "The maximum number of results to return for a single request.", isRequired = false, type = Type.INTEGER),
```
Similar to the rest docs, make sure this matches the rest of the APIs.

staubesv replied:
Fixed once https://github.com/plapadoo/opencast/pull/19 is merged

gregorydlogan commented on modules/external-api/src/main/java/org/opencastproject/external/endpoint/EventsEndpoint.java:
```diff
@@ -152,6 +164,7 @@
 import javax.ws.rs.core.Response.Status;
 
 @Path("/")
+@Produces({ "application/json", "application/v1.0.0+json", "application/v1.1.0+json" })
```
Constants?

gregorydlogan commented on modules/external-api/src/main/java/org/opencastproject/external/endpoint/GroupsEndpoint.java:
```diff
@@ -70,6 +69,7 @@
 import javax.ws.rs.core.Response;
 
 @Path("/")
+@Produces({ "application/json", "application/v1.0.0+json", "application/v1.1.0+json" })
```
Constants?

staubesv replied:
Fixed once https://github.com/plapadoo/opencast/pull/19 is merged

gregorydlogan commented on modules/external-api/src/main/java/org/opencastproject/external/endpoint/SecurityEndpoint.java:
```diff
@@ -63,6 +62,7 @@
 import javax.ws.rs.core.Response;
 
 @Path("/")
+@Produces({ "application/json", "application/v1.0.0+json", "application/v1.1.0+json" })
```
Constants?

staubesv replied:
Fixed once https://github.com/plapadoo/opencast/pull/19 is merged

gregorydlogan commented on modules/external-api/src/main/java/org/opencastproject/external/endpoint/SeriesEndpoint.java:
```diff
@@ -119,6 +119,7 @@
 import javax.ws.rs.core.Response.Status;
 
 @Path("/")
+@Produces({ "application/json", "application/v1.0.0+json", "application/v1.1.0+json" })
```
Constants?

staubesv replied:
Fixed once https://github.com/plapadoo/opencast/pull/19 is merged

gregorydlogan commented on modules/external-api/src/main/java/org/opencastproject/external/util/CaptureAgentUtils.java:
```diff
@@ -0,0 +1,70 @@
+    fields.add(f(JSON_KEY_AGENT_ID, v(agent.getName())));
+    fields.add(f(JSON_KEY_UPDATE, v(toUTC(agent.getLastHeardFrom()), Jsons.BLANK)));
+    fields.add(f(JSON_KEY_URL, v(agent.getUrl(), Jsons.BLANK)));
+    fields.add(f(JSON_KEY_INPUTS, (StringUtils.isEmpty(devices)) ? arr() : arr(devices.split(","))));
```
Suggestion: Add agent reported timezone, if any.

staubesv replied:
The timezones reported by capture agents should not be relevant to API clients as the backend is supposed to ensure the correct handling of timezones for scheduling.

gregorydlogan replied:
It might be if they're building a scheduling UI over top and need to support multiple timezones.  For instance, if I'm in UTC-6, and scheduling an agent in UTC+2, it might be helpful to *know* that.

I suppose this comes down to the bugs in the current scheduling UI re: timezones.  Perhaps we leave this alone and if we find we need it we can add it in API 1.2.0.

staubesv replied:
In case somebody wants to add (parts of) the capture agent configuration, it could be added to 1.1.0 (before r/6.x is cut) or in a future API version. 
It would require some thoughts (if done properly) as it is not that clear (at least to me) which parts of the capture agent configuration are "standard", "legacy" or "vendor" - and just adding all would unnecessarily blow up this response which again is undesirable.

Anyway, this PR makes that easier as it is now quite obvious how versioning works as the code supports multiple versions.

As for us, though, we simply don't have that requirement.

gregorydlogan commented on modules/external-api/src/main/resources/OSGI-INF/external-events-endpoint.xml:
```diff
@@ -47,4 +47,14 @@
             cardinality="1..1"
             policy="static"
             bind="setUrlSigningService" />
+  <reference name="SchedulerService"
```
Incorrect number of spaces.  Ugh, this whole file is all out of whack.  Reformat entirely?

staubesv replied:
The formatting is inconsistent not just within this file but rather across all OSGI service declarations of the module. I've created https://opencast.jira.com/browse/MH-12968 for that.

I'm willing to raise another PR that "fixes" MH-12968 (for this module), but I don't think that it makes sense to do this in this PR as this PR is actually not directly related to MH-12968.

That raises a question: How should a OSGI service declaration be formatted? (as this also affects all other modules)

gregorydlogan replied:
See https://github.com/opencast/opencast/pull/314.  I'm assuming that formatting will get rolled into this PR, at least for this file?

staubesv replied:
I would expect #314 to wait until this PR is merged as this PR was first and has been waiting for quite a while.

gregorydlogan reviewed (changes requested):
Looks really good, a few minor issues.

staubesv committed:
Unify documentation of pagination across

Although the paginations is actually implemented the same way for all endpoints of the External API, the documentation was sometimes wrong and sometimes not consistent. This commit unifies the documentation of the query parameters 'offset' and 'limit' across the External API specification and the REST Docs.

staubesv committed:
Fixed typo in documentation of the Events endpoint

staubesv committed:
Use constants in @Produces annotation

staubesv commented:
Hi @gregorydlogan 

Thanks for taking the review!

I think I've resolved all your comments (sometimes with an question or alternative opinion).

-> Your turn again.

staubesv commented on etc/security/mh_default_org.xml:
```diff
@@ -162,6 +162,8 @@
     <!-- External API GET Endpoints -->
     <sec:intercept-url pattern="/api" method="GET" access="ROLE_ADMIN, ROLE_API"/>
+    <sec:intercept-url pattern="/api/agents" method="GET" access="ROLE_ADMIN, ROLE_API_CAPTURE_AGENTS_VIEW"/>
+    <sec:intercept-url pattern="/api/agents/*" method="GET" access="ROLE_ADMIN, ROLE_API, ROLE_API_CAPTURE_AGENTS_VIEW"/>
```
@gregorydlogan But here, ROLE_API is not required. @krinnewitz will update the PR in a few moments.

krinnewitz replied:
Done ;)

Commit:
Remove unnecessary references to ROLE_API

staubesv committed:
Improve documentation of POST /api/events (#20)

* Improve documentation of POST /api/events

staubesv commented:
Hi @gregorydlogan 

Just completed another round of resolving comments: The documentation is now more precise considering the use of end vs. duration and UTC for scheduling. rrule has become a dedicated data type.

-> Your turn again.

staubesv commented:
@krinnewitz Could you please resolve the "error" I've introduced with my last PR?

./developer/docs/api/events-api.md: 183: MD009/no-trailing-spaces Trailing spaces [Expected: 0 or 2; Actual: 1]
 Use --force to continue.

Sorry for that.

krinnewitz commented:
Ok done
